### PR TITLE
Allow Umpire to be used with no detected devices

### LINKED ISSUE #712: Don't error when compiled with cuda but no device present.
**Is your feature request related to a problem? Please describe.**
We deploy the same container to VM's that might or might not have GPUs, but we only create a single image. If we don't have GPUs, we would never request device memory, and would error out before any request for a device allocator. 

**Describe the solution you'd like**

Only error if configured with cuda and device memory is requested. I think the simple solution might be to issue a warning when detecting the devices in `MemoryResourceRegistry.cpp` and don't add the device allocator. I assume this would throw an error if the `device` allocator was ever requested.

I can patch umpire, but it might be nice to have this work in heterogeneous cloud envs. Here is an example patch that works for me:
```
diff -ruN a/src/umpire/resource/MemoryResourceRegistry.cpp b/src/umpire/resource/MemoryResourceRegistry.cpp
--- a/src/umpire/resource/MemoryResourceRegistry.cpp	2022-03-10 18:30:58.008480346 +0000
+++ b/src/umpire/resource/MemoryResourceRegistry.cpp	2022-03-10 18:32:47.110041452 +0000
@@ -108,27 +108,30 @@
     int device_count{0};
     auto error = ::cudaGetDeviceCount(&device_count);
     if (error != cudaSuccess) {
-      UMPIRE_ERROR("Umpire compiled with CUDA support but no GPUs detected!");
+      //UMPIRE_WARNING("Umpire compiled with CUDA support but no GPUs detected! Don't you dare ask me for a device allocator!!!!");
     }
+    else
+    {
 
-    registerMemoryResource(util::make_unique<resource::CudaDeviceResourceFactory>());
-    m_resource_names.push_back("DEVICE");
+      registerMemoryResource(util::make_unique<resource::CudaDeviceResourceFactory>());
+      m_resource_names.push_back("DEVICE");
 
-    for (int device = 0; device < device_count; ++device) {
-      std::string name{"DEVICE::" + std::to_string(device)};
-      m_resource_names.push_back(name);
-    }
+      for (int device = 0; device < device_count; ++device) {
+        std::string name{"DEVICE::" + std::to_string(device)};
+        m_resource_names.push_back(name);
+      }
 
-    registerMemoryResource(util::make_unique<resource::CudaUnifiedMemoryResourceFactory>());
-    m_resource_names.push_back("UM");
+      registerMemoryResource(util::make_unique<resource::CudaUnifiedMemoryResourceFactory>());
+      m_resource_names.push_back("UM");
 
-    registerMemoryResource(util::make_unique<resource::CudaPinnedMemoryResourceFactory>());
-    m_resource_names.push_back("PINNED");
+      registerMemoryResource(util::make_unique<resource::CudaPinnedMemoryResourceFactory>());
+      m_resource_names.push_back("PINNED");
 
 #if defined(UMPIRE_ENABLE_CONST)
-    registerMemoryResource(util::make_unique<resource::CudaConstantMemoryResourceFactory>());
-    m_resource_names.push_back("DEVICE_CONST");
+      registerMemoryResource(util::make_unique<resource::CudaConstantMemoryResourceFactory>());
+      m_resource_names.push_back("DEVICE_CONST");
 #endif
+    }
   }
 #endif
```

### PR DESCRIPTION
**Pull request by davidbeckingsale**
Closes #712

### CODE REVIEW

File: src/umpire/resource/MemoryResourceRegistry.cpp [hunk 1]
```diff
@@ -108,27 +108,27 @@ MemoryResourceRegistry::MemoryResourceRegistry() : m_allocator_factories()
...
-    }
-
-    registerMemoryResource(util::make_unique<resource::CudaDeviceResourceFactory>());
-    m_resource_names.push_back("DEVICE");
-
-    for (int device = 0; device < device_count; ++device) {
-      std::string name{"DEVICE::" + std::to_string(device)};
-      m_resource_names.push_back(name);
-    }
+      UMPIRE_LOG(Warning, "Umpire compiled with CUDA support but no GPUs detected!");
```
Comment (mcfadden8) [inline src/umpire/resource/MemoryResourceRegistry.cpp:111]: Is there a specific error code that can be checked for device existence?  I assume that in theory, other error codes code be returned?

File: src/umpire/resource/MemoryResourceRegistry.cpp [hunk 2]
```diff
@@ -137,27 +137,27 @@ MemoryResourceRegistry::MemoryResourceRegistry() : m_allocator_factories()
...
-    }
-
-    registerMemoryResource(util::make_unique<resource::HipDeviceResourceFactory>());
-    m_resource_names.push_back("DEVICE");
-
-    for (int device = 0; device < device_count; ++device) {
-      std::string name{"DEVICE::" + std::to_string(device)};
-      m_resource_names.push_back(name);
-    }
+      UMPIRE_LOG(Warning, "Umpire compiled with HIP support but no GPUs detected!");
```
Comment (mcfadden8) [inline src/umpire/resource/MemoryResourceRegistry.cpp:140]: Same error code question as above.  Also, how is this being tested?  Has this been manually tested on corona?

File: src/umpire/resource/MemoryResourceRegistry.cpp [hunk 3]
```diff
@@ -181,34 +181,39 @@ MemoryResourceRegistry::MemoryResourceRegistry() : m_allocator_factories()
     }
 
     if (device_count == 0) {
-      UMPIRE_ERROR("Umpire compiled with SYCL support but no GPUs detected!");
-    }
+      UMPIRE_LOG(Warning, "Umpire compiled with SYCL support but no GPUs detected!");
```
Comment (mcfadden8) [inline src/umpire/resource/MemoryResourceRegistry.cpp:184]: Is this code path being tested?

**Comment by mclarsen** [comment]
Thanks!

**Pull request merged**